### PR TITLE
Secret Token Validation feature

### PR DESCRIPTION
**Pull request by barv-jfrog**
- [x] The pull request is targeting the `dev` branch.
- [x] The code has been validated to compile successfully by running `go vet ./...`.
- [x] The code has been formatted properly using `go fmt ./...`.
- [x] All [static analysis checks](https://github.com/jfrog/jfrog-cli-security/actions/workflows/analysis.yml) passed.
- [x] All [tests](https://github.com/jfrog/jfrog-cli-security/actions/workflows/test.yml) have passed. If this feature is not already covered by the tests, new tests have been added.
- [x] All changes are detailed at the description. if not already covered at [JFrog Documentation](https://github.com/jfrog/documentation), new documentation have been added.

-----
Depends on - https://github.com/jfrog/documentation/pull/145
Depends on - https://github.com/jfrog/jfrog-client-go/pull/991

-----

Description - adding a flag --validate-secrets to jf audit and jf docker scan so secrets found will trigger token validation on XRAY. Token validation takes secrets that are api tokens for example amazon secret key and checks if this key is still valid on amazon side.  The capability is identical to audit and docker scan. 
What I do is I pass an env var to analyzermanager because analyzers contain an env variable which according to its value (true/false) turns on the Gadget which is responsible for token validation.
As you see, there are multiple options to pass this env var, first through flag --validate-secrets, second through env var defined in user setup, third XRAY API which exists only from 3.101.0. otherwise it returns False.

-----

OLD PR - https://github.com/jfrog/jfrog-cli-security/pull/128

-----

image (8)

### CODE REVIEW

File: cli/docs/flags.go [hunk 1]
```diff
@@ -199,14 +200,15 @@ var flagsMap = map[string]components.Flag{
...
+	SpecFlag:         components.NewStringFlag(SpecFlag, "Path to a File Spec."),
+	scanRecursive:    components.NewBoolFlag(Recursive, "Set to false if you do not wish to collect artifacts in sub-folders to be scanned by Xray.", components.WithBoolDefaultValue(true)),
+	scanRegexp:       components.NewBoolFlag(RegexpFlag, "Set to true to use a regular expression instead of wildcards expression to collect files to scan."),
+	scanAnt:          components.NewBoolFlag(AntFlag, "Set to true to use an ant pattern instead of wildcards expression to collect files to scan."),
+	Project:          components.NewStringFlag(Project, "JFrog Artifactory project key."),
+	Watches:          components.NewStringFlag(Watches, "A comma-separated list of Xray watches, to determine Xray's violations creation."),
+	RepoPath:         components.NewStringFlag(RepoPath, "Target repo path, to enable Xray to determine watches accordingly."),
+	Licenses:         components.NewBoolFlag(Licenses, "Set to true if you'd like to receive licenses from Xray scanning."),
+	SecretValidation: components.NewBoolFlag(SecretValidation, "Set to true if you want exposures scanner to validate api tokens"),
```
Comment (attiasas) [inline cli/docs/flags.go:211]: move it below `WithoutCA`, make the text similar to it and add note about adding `--secrets`

File: utils/resultwriter.go [hunk 2]
```diff
@@ -159,6 +159,7 @@ func (rw *ResultsWriter) printScanResultsTables() (err error) {
 		if err = PrintSecretsTable(rw.results.ExtendedScanResults.SecretsScanResults, rw.results.ExtendedScanResults.EntitledForJas); err != nil {
 			return
 		}
+		log.Output("This table contains multiple secret types, such as tokens, generic password, ssh keys and more, token validation is only supported on tokens.")
```
Comment (attiasas) [inline utils/resultwriter.go:162]: log this only in case secret validation requested and jas enabled

File: utils/resultstable_test.go [hunk 3]
```diff
@@ -1097,11 +1097,77 @@ func TestPrepareSecrets(t *testing.T) {
...
+					},
+				},
+			},
+		},
 	}
 
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
-			assert.ElementsMatch(t, tc.expectedOutput, prepareSecrets(tc.input, false))
+			rows, _ := prepareSecrets(tc.input, false)
```
Comment (attiasas) [inline utils/resultstable_test.go:1169]: Assert the returned bool as well

File: utils/jasutils/jasutils.go [hunk 4]
```diff
@@ -82,6 +86,22 @@ var applicableMapToScore = map[string]int{
...
 }
 
+var TokenValidationOrder = map[string]int{
+	"Active":      1,
+	"Inactive":    2,
+	"Unsupported": 3,
+	"Unavailable": 4,
+	"":            5,
+	"Not a token": 6,
+}
```
Comment (attiasas) [inline utils/jasutils/jasutils.go:96]: Create an Enum for the keys (SecretStatus), use it instead of Const

File: formats/conversion.go [hunk 5]
```diff
@@ -141,14 +142,25 @@ func ConvertToOperationalRiskViolationScanTableRow(rows []OperationalRiskViolati
...
+		newRow := secretsTableRow{
 			severity:   rows[i].Severity,
 			file:       rows[i].File,
 			lineColumn: strconv.Itoa(rows[i].StartLine) + ":" + strconv.Itoa(rows[i].StartColumn),
 			secret:     rows[i].Snippet,
-		})
+		}
+		if tokenValidation {
+			newRow.tokenValidation = "Not a token"
+			newRow.tokenInfo = "No Extra Info"
```
Comment (attiasas) [inline formats/conversion.go:155]:
```suggestion
			newRow.tokenValidation = "Not a token"
```
Leave empty or use `N/A`, use enums string values or const

File: jas/common.go [hunk 6]
```diff
@@ -273,8 +273,27 @@ func convertToFilesExcludePatterns(excludePatterns []string) []string {
 	return patterns
 }
 
-func GetAnalyzerManagerXscEnvVars(msi string, technologies ...techutils.Technology) map[string]string {
+func CheckForSecretValidation(xrayManager *xray.XrayServicesManager, xrayVersion string, validateSecrets bool) bool {
+	dynamicTokenVersionMismatchErr := goclientutils.ValidateMinimumVersion(goclientutils.Xray, xrayVersion, jasutils.DynamicTokenValidationMinXrayVersion)
+	if dynamicTokenVersionMismatchErr != nil {
+		if validateSecrets {
+			log.Warn("Token validation (--validate-secrets flag) is not supported in your xray version")
+		}
```
Comment (attiasas) [inline jas/common.go:281]:
```suggestion
		if validateSecrets {
			log.Info("Token validation (--validate-secrets flag) is not supported in your xray version")
		}
```
add what version and what is the minimum in the log. I think you can print the error and the info is in it just wrap it with fmt and the extra info

File: jas/common.go [hunk 7]
```diff
@@ -273,8 +273,27 @@ func convertToFilesExcludePatterns(excludePatterns []string) []string {
...
+		return false
+	}
+	// Ordered By importance
+	// first check for flag and second check for env var
+	if validateSecrets || strings.ToLower(os.Getenv("JF_VALIDATE_SECRETS")) == "true" {
```
Comment (attiasas) [inline jas/common.go:286]:
```suggestion
	if validateSecrets || strings.ToLower(os.Getenv(JfSecretValidationEnvVariable)) == "true" {
```
Use the const you defined...

File: jas/common_test.go [hunk 8]
```diff
@@ -183,7 +183,7 @@ func TestGetAnalyzerManagerXscEnvVars(t *testing.T) {
 	}
 	for _, test := range tests {
 		t.Run(test.name, func(t *testing.T) {
-			assert.Equal(t, test.expectedOutput, GetAnalyzerManagerXscEnvVars(test.msi, test.technologies...))
+			assert.Equal(t, test.expectedOutput, GetAnalyzerManagerXscEnvVars(test.msi, false, test.technologies...))
```
Comment (attiasas) [inline jas/common_test.go:186]: Add test cases for `true` and `false`

File: utils/resultstable.go [hunk 9]
```diff
@@ -339,11 +339,13 @@ func PrepareLicenses(licenses []services.License) ([]formats.LicenseRow, error)
 
 // Prepare secrets for all non-table formats (without style or emoji)
 func PrepareSecrets(secrets []*sarif.Run) []formats.SourceCodeRow {
-	return prepareSecrets(secrets, false)
+	preparedSecrets, _ := prepareSecrets(secrets, false)
```
Comment (attiasas) [inline utils/resultstable.go:342]: Used for SimpleJson, make sure the values are also added to this format or you need to pass the bool

File: utils/resultstable.go [hunk 10]
```diff
@@ -352,36 +354,43 @@ func prepareSecrets(secrets []*sarif.Run, isTable bool) []formats.SourceCodeRow
...
+				secretsRows = append(secretsRows, newRow)
 			}
 		}
 	}
 
 	sort.Slice(secretsRows, func(i, j int) bool {
+		if jasutils.TokenValidationOrder[secretsRows[i].TokenValidation] != jasutils.TokenValidationOrder[secretsRows[j].TokenValidation] {
+			return jasutils.TokenValidationOrder[secretsRows[i].TokenValidation] < jasutils.TokenValidationOrder[secretsRows[j].TokenValidation]
+		}
 		return secretsRows[i].SeverityNumValue > secretsRows[j].SeverityNumValue
```
Comment (attiasas) [inline utils/resultstable.go:383]: I think you need more complex sorting like with "ApplicableStatus" + Severity.
If the status is `Active` or `Unsupported` or `Not scanned` it should be base on severity and above the `not active tokens`.
Right now the order does not make sense

Comment (attiasas) [reply utils/resultstable.go:383]: add tests for the sorting as well

File: tests/utils/test_validation.go [hunk 11]
```diff
@@ -65,6 +65,34 @@ func VerifySimpleJsonScanResults(t *testing.T, content string, minViolations, mi
...
+		count_inactives := 0
+		for _, result := range results.Secrets {
+			if result.TokenValidation == "Inactive" {
+				count_inactives += 1
+			}
+		}
+		assert.GreaterOrEqual(t, count_inactives, minInactives)
+	}
+}
+
```
Comment (attiasas) [inline tests/utils/test_validation.go:95]: why only "Incative"? what about the other statuses? why not add it to the already existing VerifySimpleJson funcs?...

File: scans_test.go [hunk 12]
```diff
@@ -161,6 +166,22 @@ func runDockerScan(t *testing.T, testCli *coreTests.JfrogCli, imageName, watchNa
...
+		defer commonTests.DeleteTestImage(t, imageTag, containerUtils.DockerClient)
+		// Run docker scan on image
+		cmdArgs := []string{"docker", "scan", imageTag, "--validate-secrets", "--server-id=default", "--licenses", "--format=simple-json", "--fail=false", "--min-severity=low", "--fixable-only"}
+		output := testCli.WithoutCredentials().RunCliCmdWithOutput(t, cmdArgs...)
+		if assert.NotEmpty(t, output) {
+			securityTestUtils.VerifyJsonScanTokenValidationResults(t, output, minInactives)
+		}
+	}
+}
+
```
Comment (attiasas) [inline scans_test.go:184]: similar code to `runDockerScan` can you add an option there?

**Review by attiasas — CHANGES_REQUESTED** [review]

File: audit_test.go [hunk 13]
```diff
@@ -505,9 +511,17 @@ func testXrayAuditJas(t *testing.T, testCli *coreTests.JfrogCli, project string,
 	assert.NoError(t, err)
 	chdirCallback := clientTests.ChangeDirWithCallback(t, baseWd, tempDirPath)
 	defer chdirCallback()
+	if validateSecrets[0] {
+		return testCli.WithoutCredentials().RunCliCmdWithOutput(t, "audit", "--validate-secrets", "--secrets", "--format="+string(format.SimpleJson), "--threads="+threads)
+	}
 	return testCli.WithoutCredentials().RunCliCmdWithOutput(t, "audit", "--format="+string(format.SimpleJson), "--threads="+threads)
```
[the remark this answers is not in the document]
Comment (barv-jfrog) [reply audit_test.go:517]: fixed

File: cli/docs/flags.go (hunk 1 above)
Comment (barv-jfrog) [reply cli/docs/flags.go:211]: done

File: formats/conversion.go (hunk 5 above)
Comment (barv-jfrog) [reply formats/conversion.go:155]: This texts were required by Eyal and Asaf Karas. I will move it to an enum.

File: audit_test.go [hunk 14]
```diff
@@ -491,7 +497,7 @@ func TestXrayAuditJasNoViolationsSimpleJson(t *testing.T) {
 	securityTestUtils.VerifySimpleJsonJasResults(t, output, 0, 0, 0, 0, 0, 0, 1)
 }
 
-func testXrayAuditJas(t *testing.T, testCli *coreTests.JfrogCli, project string, threads string) string {
+func testXrayAuditJas(t *testing.T, testCli *coreTests.JfrogCli, project string, threads string, validateSecrets ...bool) string {
```
[the remark this answers is not in the document]
Comment (barv-jfrog) [reply audit_test.go:500]: fixed

File: jas/common.go (hunk 6 above)
Comment (barv-jfrog) [reply jas/common.go:281]: ok

File: jas/common.go (hunk 7 above)
Comment (barv-jfrog) [reply jas/common.go:286]: right

File: jas/common_test.go (hunk 8 above)
Comment (barv-jfrog) [reply jas/common_test.go:186]: ok

File: scans_test.go (hunk 12 above)
Comment (barv-jfrog) [reply scans_test.go:184]: yes

File: tests/utils/test_validation.go (hunk 11 above)
Comment (barv-jfrog) [reply tests/utils/test_validation.go:95]: Only inactive because active requires me to push active tokens which is obviously not possible.

File: utils/jasutils/jasutils.go (hunk 4 above)
Comment (barv-jfrog) [reply utils/jasutils/jasutils.go:96]: done

File: utils/resultstable_test.go (hunk 3 above)
Comment (barv-jfrog) [reply utils/resultstable_test.go:1169]: ok

File: utils/resultwriter.go (hunk 2 above)
Comment (barv-jfrog) [reply utils/resultwriter.go:162]: right

File: utils/resultstable.go (hunk 10 above)
Comment (barv-jfrog) [reply utils/resultstable.go:383]: we discussed about this, also added tests for the sorting

File: utils/resultstable.go (hunk 9 above)
Comment (barv-jfrog) [reply utils/resultstable.go:342]: simplejson works well

File: go.mod [hunk 15]
```diff
@@ -116,7 +116,7 @@ require (
 
 // replace github.com/jfrog/jfrog-cli-core/v2 => github.com/jfrog/jfrog-cli-core/v2 dev
 
-// replace github.com/jfrog/jfrog-client-go => github.com/jfrog/jfrog-client-go dev
+//replace github.com/jfrog/jfrog-client-go => github.com/jfrog/jfrog-client-go dev
```
Comment (attiasas) [inline go.mod:119]: revert this change

File: scans_test.go [hunk 16]
```diff
@@ -118,15 +118,20 @@ func TestDockerScan(t *testing.T) {
...
 	securityTestUtils.ValidateXrayVersion(t, "3.41.0")
 
 	// Image with 0 vulnerabilities
-	runDockerScan(t, testCli, "busybox:1.35", "", 0, 0, 0)
+	runDockerScan(t, testCli, "busybox:1.35", "", 0, 0, 0, 0, false)
+
+	securityTestUtils.InitSecurityTest(t, jasutils.DynamicTokenValidationMinXrayVersion)
+	// #nosec G101 -- Image with dummy token for tests
+	tokensImageToScan := "srmishj/inactive_tokens:latest"
+	runDockerScan(t, testCli, tokensImageToScan, "", 0, 0, 0, 5, true)
```
[the remark this answers is not in the document]
Comment (barv-jfrog) [reply scans_test.go:134]: ok

File: utils/jasutils/jasutils.go [hunk 17]
```diff
@@ -82,6 +113,15 @@ var applicableMapToScore = map[string]int{
...
 }
 
+var TokenValidationOrder = map[string]int{
+	"Active":      1,
+	"Unsupported": 2,
+	"Unavailable": 3,
+	"Inactive":    4,
+	"Not a token": 5,
+	"":            6,
+}
```
[the remark this answers is not in the document]
Comment (barv-jfrog) [reply utils/jasutils/jasutils.go:123]: I'm not sure. I believe what I did is fine, since I have ToString(), it has similar functionality like ApplicabilityStatus, don't you agree?

File: formats/conversion.go [hunk 18]
```diff
@@ -141,14 +142,25 @@ func ConvertToOperationalRiskViolationScanTableRow(rows []OperationalRiskViolati
...
+			newRow.tokenValidation = jasutils.TokenValidationStatusForNonTokens
+			newRow.tokenInfo = jasutils.MissingMetadata
+			if rows[i].TokenValidation != "" {
+				newRow.tokenValidation = jasutils.TokenValidationStatus(rows[i].TokenValidation).ToString()
+			}
+			if rows[i].TokenInfo != "" {
+				newRow.tokenInfo = rows[i].TokenInfo
+			}
+		}
+		tableRows = append(tableRows, newRow)
```
[the remark this answers is not in the document]
Comment (barv-jfrog) [reply formats/conversion.go]: we don't

File: tests/utils/test_validation.go [hunk 19]
```diff
@@ -55,13 +55,20 @@ func VerifyJsonScanResults(t *testing.T, content string, minViolations, minVulne
 	}
 }
 
-func VerifySimpleJsonScanResults(t *testing.T, content string, minViolations, minVulnerabilities, minLicenses int) {
+func VerifySimpleJsonScanResults(t *testing.T, content string, minViolations, minVulnerabilities, minLicenses, minInactives int) {
```
[the remark this answers is not in the document]
Comment (barv-jfrog) [reply tests/utils/test_validation.go:58]: ok

File: utils/resultstable_test.go [hunk 20]
```diff
@@ -1097,11 +1098,83 @@ func TestPrepareSecrets(t *testing.T) {
...
+				{
+					SeverityDetails: formats.SeverityDetails{
+						Severity:         "Low",
+						SeverityNumValue: 11,
+					},
+					TokenValidation: "",
+					TokenInfo:       "",
+					Finding:         "secret finding",
+					Location: formats.Location{
+						File:        "wd/file",
```
[the remark this answers is not in the document]
Comment (barv-jfrog) [reply utils/resultstable_test.go]: ok

**Review by attiasas — APPROVED** [review]
